### PR TITLE
Fix Net.Tcp duplex streaming with authentication

### LINKED ISSUE #1443: System.Xml.XmlException "The input source is not correctly formatted" when streaming over NetTcp with transport security
This is reproducible within a few seconds of single-threaded stress run.
Here is the call stack:
~~~
0:013> !PrintException /d 000001f92371cfc8
Exception object: 000001f92371cfc8
Exception type:   System.Xml.XmlException
Message:          The input source is not correctly formatted.
InnerException:   <none>

StackTrace (generated):
System_Private_DataContractSerialization!System.Xml.XmlExceptionHelper.ThrowXmlException(System.Xml.XmlDictionaryReader, System.String, System.String, System.String, System.String)+0x129
System_Private_DataContractSerialization!System.Xml.XmlBufferReader.ReadValue(System.Xml.XmlBinaryNodeType, System.Xml.ValueHandle)+0x64d
System_Private_DataContractSerialization!System.Xml.XmlBinaryReader.ReadNode()+0x145c
System_Private_ServiceModel!System.ServiceModel.Dispatcher.StreamFormatter+MessageBodyStream+<ExhaustAsync>d__14.MoveNext()+0x106
System_Private_CoreLib_ni!System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)+0x33766c
System_Private_CoreLib_ni!System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)+0x46
System_Private_ServiceModel!System.ServiceModel.Dispatcher.StreamFormatter+MessageBodyStream+<ReadAsync>d__10.MoveNext()+0x587
~~~

### PR DESCRIPTION
**Pull request by mconnew**
Fixes #1443 
Also improves performance with HTTP streaming and stops threadpool exhaustion with synchronous calls.
Also fixes an issue where our pre-fetch GET for authentication was causing a null-ref exception which was (correctly) being caught and ignored.

### CODE REVIEW

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedWriteStream.cs [hunk 1]
```diff
@@ -34,21 +33,25 @@ namespace System.ServiceModel.Channels
     /// </summary>
     internal sealed class BufferedWriteStream : Stream
     {
-        public const int DefaultBufferSize = 4096;
+        public const int DefaultBufferSize = 8192;
```
Comment (roncain) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedWriteStream.cs]: >8192 [](start = 45, length = 4)

Non-blocking question only: why this change?  Perf improvement?  Is there any embedded dependency on this being 8K?

File: src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/StreamFormatter.cs [hunk 2]
```diff
@@ -399,19 +350,19 @@ private static void Exhaust(XmlDictionaryReader reader)
...
+            //    {
+            //        do
+            //        {
+            //            if (readAheadStream != null)
+            //            {
+            //                await readAheadStream.EnsureBufferedAsync(cancellationToken);
+            //            }
+            //        } while (reader.Read());
+            //    }
+            //}
```
Comment (roncain) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/StreamFormatter.cs]: Did you intend to remove this code?

File: src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/StreamingTests.4.1.0.cs [hunk 3]
```diff
@@ -223,6 +223,81 @@ public static void NetTcp_TransportSecurity_Streamed_RoundTrips_String()
...
+            binding = new NetTcpBinding(SecurityMode.Transport);
+            binding.TransferMode = TransferMode.Streamed;
+            factory = new ChannelFactory<IWcfService>(binding, new EndpointAddress(Endpoints.Tcp_Transport_Security_Streamed_Address));
+            serviceProxy = factory.CreateChannel();
+            stream = StringToStream(testString);
+
+            // *** EXECUTE *** \\
+            var returnStream = serviceProxy.EchoStream(stream);
+            var ms = new MemoryStream((int)stream.Length);
+            var buffer = new byte[10];
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/StreamingTests.4.1.0.cs]: >10 [](start = 34, length = 2)

Just curious: did you choose a tiny buffer just to hammer the ReadAsync calls?  Is there any advantage to using a number 20000 is not divisible by to verify what happens on a partial read of < buffer.Length?

File: src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/StreamingTests.4.1.0.cs [hunk 4]
```diff
@@ -223,6 +223,81 @@ public static void NetTcp_TransportSecurity_Streamed_RoundTrips_String()
...
+            Console.WriteLine("---- Test SKIPPED --------------");
+            Console.WriteLine("Attempting to run the test in ToF, a ConditionalFact evaluated as FALSE.");
+            Console.WriteLine("Root_Certificate_Installed evaluated as {0}", root_Certificate_Installed);
+            Console.WriteLine("Client_Certificate_Installed evaluated as {0}", client_Certificate_Installed);
+            Console.WriteLine("Windows_Authentication_Available evaluated as {0}", windows_Authentication_Available);
+            Console.WriteLine("Ambient_Credentials_Available evaluated as {0}", ambient_Credentials_Available);
+            return;
+        }
+#endif
+        string testString = new string('a', 20000);
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/StreamingTests.4.1.0.cs]: >new string('a', 20000) [](start = 28, length = 22)

Non-blocking nit: technically the ReadStreamAsync could return buffer[0] 20000 times and this test would still pass.  Does it make sense to use ScenarioTestHelpers.InterestingString to mix things up a bit?

File: src/System.Private.ServiceModel/src/Internals/System/Runtime/TaskHelpers.cs [hunk 5]
```diff
@@ -311,8 +337,8 @@ public bool IsCompleted
             get
             {
                 return (TaskScheduler.Current == TaskScheduler.Default) &&
-                       (SynchronizationContext.Current == null ||
-                       (SynchronizationContext.Current.GetType() == typeof(SynchronizationContext)));
+                    (SynchronizationContext.Current == null ||
+                    (SynchronizationContext.Current.GetType() == typeof(SynchronizationContext)));
```
Comment (roncain) [inline src/System.Private.ServiceModel/src/Internals/System/Runtime/TaskHelpers.cs]: >GetType() [](start = 52, length = 9)

Just checking -- does this work okay in Net Native?

File: src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs [hunk 6]
```diff
@@ -0,0 +1,664 @@
...
+
+            // This method is only valid if you already know that (gate & HiBits) != 0.
+            public static bool IsComplete(int gate)
+            {
+                return (gate & HiMask) == gate << HiShift;
+            }
+        }
+
+        static IOThreadScheduler current = new IOThreadScheduler(32, 32);
+        //readonly ScheduledOverlapped overlapped;
```
Comment (roncain) [inline src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs]: >//readonly ScheduledOverlapped overlapped [](start = 8, length = 41)

Remove?

File: src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs [hunk 7]
```diff
@@ -0,0 +1,664 @@
...
+
+        readonly Slot[] slots;
```
Comment (roncain) [inline src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs]: >readonly Slot[] slots; [](start = 8, length = 22)

I think these need to use "private" and use underscore naming convention.

File: src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs [hunk 8]
```diff
@@ -0,0 +1,664 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics.Contracts;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Runtime
+{
+    internal class IOThreadScheduler
```
Comment (roncain) [inline src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs]: >IOThreadScheduler [](start = 19, length = 17)

Could you give us a quick overview of where this code came from and what parts are verbatim copies?  How much should I actually review?

File: src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs [hunk 9]
```diff
@@ -0,0 +1,664 @@
...
+        // Therefore, we are passing the reference, when we post a pending callback and reset it, once the callback was
+        // invoked; during that time the scheduler is rooted but in that time we don't want that it would be collected
+        // by the GC anyway.
+        //[Fx.Tag.SecurityNote(Critical = "manages NativeOverlapped instance, can be called outside user context")]
+        //[SecurityCritical]
+        //unsafe class ScheduledOverlapped
+        //{
+        //    readonly NativeOverlapped* nativeOverlapped;
+        //    IOThreadScheduler scheduler;
+
```
Comment (roncain) [inline src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs]: Remove this commented out code?

**Comment by roncain** [comment]
LGTM
There are some non-blocking questions and some nit-level code cleanup I'd like you to consider.

File: src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs [hunk 10]
```diff
@@ -0,0 +1,664 @@
...
+            Contract.Assert(capacityLowPri > 0, "Low-priority capacity must be positive.");
+            Contract.Assert(capacityLowPri <= 0x8000, "Low-priority capacity cannot exceed 32k.");
+
+            slots = new Slot[capacity];
+            Contract.Assert((slots.Length & SlotMask) == 0, "Capacity must be a power of two.");
+
+            slotsLowPri = new Slot[capacityLowPri];
+            Contract.Assert((slotsLowPri.Length & SlotMaskLowPri) == 0, "Low-priority capacity must be a power of two.");
+
+            //overlapped = new ScheduledOverlapped();
```
Comment (iamjasonp) [inline src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs]: NIT: Should this be removed?

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedWriteStream.cs [hunk 11]
```diff
@@ -57,12 +60,12 @@ public BufferedWriteStream(Stream stream, int bufferSize)
...
-                throw new ObjectDisposedException("BufferedWriteStream");
+                throw new ObjectDisposedException(nameof(BufferedWriteStream));
         }
 
         private void EnsureCanWrite()
         {
-            Contract.Requires(_stream != null);
+            Contract.Assert(_stream != null);
             if (!_stream.CanWrite)
                 throw new NotSupportedException("write");
```
Comment (iamjasonp) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedWriteStream.cs]: I know this message shouldn't happen... that said, should we fix the exception message here?

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedReadStream.cs [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+        }
+
+        public override void WriteByte(byte value)
+        {
+            throw new NotSupportedException("WriteByte");
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException("seek");
```
Comment (iamjasonp) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedReadStream.cs]: NIT: "Seek" capitalization

Comment (iamjasonp) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedReadStream.cs]: Or better yet, use `nameof` ?

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedReadStream.cs [hunk 13]
```diff
@@ -0,0 +1,444 @@
...
+        }
+
+        public override bool CanSeek
+        {
+            get { return false; }
+        }
+
+        public override long Length
+        {
+            get { throw new NotSupportedException("Position"); }
```
Comment (iamjasonp) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedReadStream.cs]: Should be "Length" or `nameof(Length)`

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedWriteStream.cs (hunk 1 above)
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedWriteStream.cs]: Yes, perf improvement. As we have a synchronous read/write path in streaming scenarios, this reduces the number of times we require 2 threads to do anything. Additionally, the buffers now comes from the BufferManager so if the stream is disposed, this will reduce memory consumption so the larger buffer won't have a negative effect.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/StreamFormatter.cs (hunk 2 above)
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/StreamFormatter.cs]: Yes, I'll remove it.

File: src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/StreamingTests.4.1.0.cs (hunk 3 above)
Comment (mconnew) [inline src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Tcp/StreamingTests.4.1.0.cs]: Good idea, I'll make it 20001.

File: src/System.Private.ServiceModel/src/Internals/System/Runtime/TaskHelpers.cs (hunk 5 above)
Comment (mconnew) [inline src/System.Private.ServiceModel/src/Internals/System/Runtime/TaskHelpers.cs]: This is existing code, looks like it got reformatted. This works fine in Net Native as this is easily discoverable by the analysis engine. Also, I'm not needing runtime types so I don't think even reflection information is required.

File: src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs (hunk 7 above)
Comment (mconnew) [inline src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs]: Yep, copied from full framework code and forgot to apply new formatting.

File: src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs (hunk 8 above)
Comment (mconnew) [inline src/System.Private.ServiceModel/src/Internals/System/Runtime/IOThreadScheduler.cs]: Came from full framework class of the same name. Everything is identical except how we launch a thread to process the work items. Instead of posting a callback to another thread using IOOverlapped's I'm using TaskScheduler managed threads which aren't part of the ThreadPool. Basically the lines which say `Task.Factory.StartNew` used to call a different method to create a worker thread.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedWriteStream.cs (hunk 11 above)
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedWriteStream.cs]: I would prefer to change this to a Contract.Assert so as not to run the code on release builds. The original code was general use and any stream could be passed in, but this implementation is only used by WCF code so the stream isn't arbitrary and a debug assert should be sufficient.

**Comment by mconnew** [comment]
@dotnet-bot test outerloop Ubuntu14.04

**Comment by mconnew** [comment]
@dotnet-bot test outerloop Ubuntu14.04

**Comment by mconnew** [comment]
@dotnet-bot test Innerloop CentOS7.1 Release

**Comment by mconnew** [comment]
@dotnet-bot test outerloop Ubuntu14.04

**Comment by iamjasonp** [comment]
LGTM
